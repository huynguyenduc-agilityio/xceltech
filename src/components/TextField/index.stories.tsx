import { useForm } from 'react-hook-form';
import { Meta, StoryObj } from '@storybook/react';

import { Form, FormField } from '../common';
import TextField from '.';

const defaultProps = {
  label: 'Email',
  placeholder: 'Please enter email',
  onChange: () => {},
};

const meta: Meta<typeof TextField> = {
  title: 'Components/TextField',
  component: TextField,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  decorators: [
    (Story) => {
      const form = useForm();
      return (
        <Form {...form}>
          <form className="flex flex-col mt-[34px] gap-2.5">
            <FormField
              control={form.control}
              name="email"
              render={({ field }) => <Story {...field} />}
            />
          </form>
        </Form>
      );
    },
  ],
} satisfies Meta<typeof TextField>;

export default meta;

type Story = StoryObj<typeof TextField>;

export const Default: Story = {
  args: { ...defaultProps },
};

export const Secondary: Story = {
  args: { ...defaultProps, variant: 'secondary' },
};

export const WithError: Story = {
  args: {
    ...defaultProps,
    errorMessage: 'This field is required',
  },
};
