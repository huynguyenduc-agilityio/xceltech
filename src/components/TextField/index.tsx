import {
  ChangeEvent,
  forwardRef,
  InputHTMLAttributes,
  useCallback,
} from 'react';
import { VariantProps } from 'class-variance-authority';

// Component
import {
  FormControl,
  FormItem,
  FormMessage,
  Input,
  inputVariants,
  Label,
} from '@/components/common';

type TTextFieldProps = InputHTMLAttributes<HTMLInputElement> &
  VariantProps<typeof inputVariants> & {
    label: string;
    errorMessage?: string;
    inputClassName?: string;
    labelClassName?: string;
    onChange: (value: string) => void;
  };

const TextField = forwardRef<HTMLInputElement, TTextFieldProps>(
  (
    {
      label,
      variant = 'primary',
      errorMessage,
      className,
      inputClassName,
      labelClassName,
      onChange,
      ...rest
    },
    ref,
  ) => {
    const handleChangeValue = useCallback(
      (e: ChangeEvent<HTMLInputElement>) => {
        onChange(e.target.value);
      },
      [onChange],
    );

    return (
      <FormItem className={className}>
        <Label
          id={`${label}-label`}
          variant={variant}
          size={variant}
          className={labelClassName}
        >
          {label}
        </Label>
        <FormControl>
          <Input
            ref={ref}
            aria-labelledby={`${label}-label`}
            variant={variant}
            className={inputClassName}
            onChange={handleChangeValue}
            {...rest}
            isInvalid={!!errorMessage}
          />
        </FormControl>
        {errorMessage && <FormMessage>{errorMessage}</FormMessage>}
      </FormItem>
    );
  },
);

export default TextField;
