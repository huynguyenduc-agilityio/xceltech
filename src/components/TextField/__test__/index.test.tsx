import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useForm, FormProvider } from 'react-hook-form';
import TextField from '..';

const Wrapper = ({ children }: { children: React.ReactNode }) => {
  const methods = useForm();

  return <FormProvider {...methods}>{children}</FormProvider>;
};

describe('TextField Component', () => {
  const mockOnChange = jest.fn();

  const commonProps = {
    label: 'Username',
    placeholder: 'Enter your username',
    onChange: mockOnChange,
  };

  it('renders the component with label and input field', () => {
    render(
      <Wrapper>
        <TextField {...commonProps} />
      </Wrapper>,
    );

    expect(screen.getByLabelText(/username/i)).toBeInTheDocument();
    expect(
      screen.getByPlaceholderText(/enter your username/i),
    ).toBeInTheDocument();
  });

  it('calls onChange when input value changes', async () => {
    render(
      <Wrapper>
        <TextField {...commonProps} />
      </Wrapper>,
    );

    const input = screen.getByLabelText(/username/i);

    await userEvent.type(input, 'Hello');

    expect(mockOnChange).toHaveBeenCalledTimes(5);
    expect(mockOnChange).toHaveBeenLastCalledWith('Hello');
  });

  it('shows error message when errorMessage is provided', () => {
    render(
      <Wrapper>
        <TextField {...commonProps} errorMessage="Invalid input" />
      </Wrapper>,
    );

    expect(screen.getByText(/invalid input/i)).toBeInTheDocument();
    expect(screen.getByText(/invalid input/i)).toHaveClass('text-red-500');
  });

  it('does not show error message when errorMessage is not provided', () => {
    render(
      <Wrapper>
        <TextField {...commonProps} />
      </Wrapper>,
    );

    expect(screen.queryByText(/invalid input/i)).not.toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const { asFragment } = render(
      <Wrapper>
        <TextField {...commonProps} />
      </Wrapper>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
