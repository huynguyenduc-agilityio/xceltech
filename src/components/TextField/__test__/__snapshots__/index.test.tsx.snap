// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="space-y-2 w-full"
  >
    <label
      class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-black-soft text-xl"
      id="Username-label"
    >
      Username
    </label>
    <input
      aria-describedby=":r4:-form-item-description"
      aria-invalid="false"
      aria-labelledby="Username-label"
      autocomplete="off"
      class="font-productsans flex w-full px-5 py-4 text-lg text-black-smoky placeholder:text-gray-soft outline-none disabled:cursor-not-allowed disabled:opacity-50 bg-blue-light h-[54px]"
      data-testid="input-element"
      id=":r4:-form-item"
      placeholder="Enter your username"
    />
  </div>
</DocumentFragment>
`;
